File: impl/app/android/app/src/test/java/com/ak/contexta/domain/usecase/StartupOrchestrationUseCaseTest.kt
package com.ak.contexta.domain.usecase

import com.ak.contexta.domain.BackgroundWorkScheduler
import com.ak.contexta.domain.model.ArticleBatch
import com.ak.contexta.domain.model.BatchStatus
import com.ak.contexta.domain.model.UserSettings
import com.ak.contexta.domain.repository.ArticleRepository
import com.ak.contexta.domain.repository.SettingsRepository
import com.ak.contexta.domain.time.TimeProvider
import io.mockk.coEvery
import io.mockk.coVerify
import io.mockk.mockk
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

class StartupOrchestrationUseCaseTest {

    private val articleRepository: ArticleRepository = mockk()
    private val settingsRepository: SettingsRepository = mockk()
    private val timeProvider: TimeProvider = mockk()
    private val triggerNextBatch: TriggerNextBatchUseCase = mockk()
    private val generationScheduler: BackgroundWorkScheduler = mockk()
    private val resendPendingAlerts: ResendPendingAlertsUseCase = mockk()

    private lateinit var useCase: StartupOrchestrationUseCase

    private val settings = UserSettings(
        isOnboarded = true,
        difficultyLevel = "LOW",
        dailyArticleCount = 5
    )

    @Before
    fun setUp() {
        useCase = StartupOrchestrationUseCase(
            articleRepository = articleRepository,
            settingsRepository = settingsRepository,
            timeProvider = timeProvider,
            triggerNextBatch = triggerNextBatch,
            generationScheduler = generationScheduler,
            resendPendingAlerts = resendPendingAlerts
        )
        coEvery { settingsRepository.getSettings() } returns settings
        coEvery { timeProvider.todayDateString() } returns "2026-08-01"
        coEvery { articleRepository.isPipelineBlocked() } returns false
        coEvery { articleRepository.getGeneratingBatches() } returns emptyList()
        coEvery { articleRepository.reconcileOrphanArticles() } returns Unit
        coEvery { triggerNextBatch.invoke(any(), any()) } returns Unit
        coEvery { generationScheduler.scheduleBatchGeneration(any(), any()) } returns true
        coEvery { resendPendingAlerts.invoke() } returns Unit
    }

    // ─── 修复核心：分配批次时传 maxRefDate，不回头选旧 seed 批次 ───────────

    @Test
    fun `未分配时 findNextReadyBatch 收到 maxRefDate 而不是 null`() = runTest {
        coEvery { articleRepository.getAssignedBatchForDate("2026-08-01") } returns null
        coEvery { articleRepository.getMaxRefBatchDate() } returns "2026-03-29"
        coEvery { articleRepository.findNextReadyBatch(any(), any()) } returns null

        useCase(currentVersionCode = 1)

        // 关键断言：afterDate 必须是 maxRefDate（严格晚于已消费批次），否则会回头选 seed 旧批次
        coVerify(exactly = 1) {
            articleRepository.findNextReadyBatch(difficulty = "LOW", afterDate = "2026-03-29")
        }
    }

    @Test
    fun `maxRefDate 之后有 READY 批次时分配给今天并触发前置生成`() = runTest {
        val nextBatch = ArticleBatch(
            id = 6,
            status = BatchStatus.READY,
            difficultyLevelSnapshot = "LOW",
            generatedOn = "2026-07-31",
            lastUpdatedAt = "2026-07-31T21:25:42+08:00"
        )
        coEvery { articleRepository.getAssignedBatchForDate("2026-08-01") } returns null
        coEvery { articleRepository.getMaxRefBatchDate() } returns "2026-03-29"
        coEvery { articleRepository.findNextReadyBatch("LOW", "2026-03-29") } returns nextBatch
        coEvery { articleRepository.assignBatchForToday(6, "2026-07-31", 5) } returns true

        val result = useCase(currentVersionCode = 1)

        assertEquals(StartupOrchestrationUseCase.StartupResult.Ready, result)
        coVerify(exactly = 1) {
            articleRepository.assignBatchForToday(batchId = 6, refBatchDate = "2026-07-31", dailyCount = 5)
        }
        coVerify(exactly = 1) { triggerNextBatch.invoke("LOW", 5) }
    }

    @Test
    fun `今天已分配时只触发前置生成 不重新查找批次`() = runTest {
        val todayBatch = ArticleBatch(
            id = 1,
            status = BatchStatus.READY,
            difficultyLevelSnapshot = "LOW",
            generatedOn = "2026-03-29",
            lastUpdatedAt = "2026-03-29T12:00:00+08:00"
        )
        coEvery { articleRepository.getAssignedBatchForDate("2026-08-01") } returns todayBatch

        val result = useCase(currentVersionCode = 1)

        assertEquals(StartupOrchestrationUseCase.StartupResult.Ready, result)
        coVerify(exactly = 1) { triggerNextBatch.invoke("LOW", 5) }
        coVerify(exactly = 0) { articleRepository.findNextReadyBatch(any(), any()) }
    }

    // ─── 无可用批次 → NeedsInitialBatch ─────────────────────────────────

    @Test
    fun `maxRefDate 之后无 READY 批次时返回 NeedsInitialBatch`() = runTest {
        coEvery { articleRepository.getAssignedBatchForDate("2026-08-01") } returns null
        coEvery { articleRepository.getMaxRefBatchDate() } returns "2026-07-31"
        coEvery { articleRepository.findNextReadyBatch("LOW", "2026-07-31") } returns null

        val result = useCase(currentVersionCode = 1)

        assertEquals(
            StartupOrchestrationUseCase.StartupResult.NeedsInitialBatch(
                difficulty = "LOW",
                dailyCount = 5
            ),
            result
        )
        coVerify(exactly = 0) { triggerNextBatch.invoke(any(), any()) }
    }

    // ─── 卡死批次重新调度 ───────────────────────────────────────────────

    @Test
    fun `GENERATING 批次在 reconcile 后被重新调度`() = runTest {
        val stuck = ArticleBatch(
            id = 6,
            status = BatchStatus.GENERATING,
            difficultyLevelSnapshot = "LOW",
            generatedOn = "2026-07-31",
            lastUpdatedAt = "2026-07-31T21:25:29+08:00"
        )
        coEvery { articleRepository.getGeneratingBatches() } returns listOf(stuck)
        coEvery { articleRepository.getAssignedBatchForDate("2026-08-01") } returns null
        coEvery { articleRepository.getMaxRefBatchDate() } returns "2026-07-31"
        coEvery { articleRepository.findNextReadyBatch("LOW", "2026-07-31") } returns null

        useCase(currentVersionCode = 1)

        coVerify(exactly = 1) { generationScheduler.scheduleBatchGeneration(6) }
    }

    // ─── 未送达告警补发 ─────────────────────────────────────────────────

    @Test
    fun `启动时补发未送达的飞书告警`() = runTest {
        coEvery { articleRepository.getAssignedBatchForDate("2026-08-01") } returns null
        coEvery { articleRepository.getMaxRefBatchDate() } returns "2026-07-31"
        coEvery { articleRepository.findNextReadyBatch("LOW", "2026-07-31") } returns null

        useCase(currentVersionCode = 1)

        coVerify(exactly = 1) { resendPendingAlerts.invoke() }
    }

    @Test
    fun `补发告警失败不影响启动主流程`() = runTest {
        coEvery { resendPendingAlerts.invoke() } throws RuntimeException("webhook down")
        coEvery { articleRepository.getAssignedBatchForDate("2026-08-01") } returns null
        coEvery { articleRepository.getMaxRefBatchDate() } returns "2026-07-31"
        coEvery { articleRepository.findNextReadyBatch("LOW", "2026-07-31") } returns null

        val result = useCase(currentVersionCode = 1)

        // 补发失败被 runCatching 吞掉，启动编排照常完成
        assertEquals(
            StartupOrchestrationUseCase.StartupResult.NeedsInitialBatch("LOW", 5),
            result
        )
    }

    // ─── 管道阻塞 ────────────────────────────────────────────────────────

    @Test
    fun `管道阻塞且版本未更新时返回 PipelineBlocked`() = runTest {
        coEvery { articleRepository.isPipelineBlocked() } returns true
        coEvery { articleRepository.recoverIfNewerVersion(1) } returns false

        val result = useCase(currentVersionCode = 1)

        assertTrue(result is StartupOrchestrationUseCase.StartupResult.PipelineBlocked)
    }
}

File: impl/app/flutter/test/domain/usecase/startup_orchestration_usecase_test.dart
import 'package:drift/native.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:contexta/data/local/database.dart';
import 'package:contexta/data/local/daos/article_daos.dart';
import 'package:contexta/data/sync/sync_articles_usecase.dart';
import 'package:contexta/domain/model/article_batch.dart';
import 'package:contexta/domain/model/user_settings.dart';
import 'package:contexta/domain/repository/article_repository.dart';
import 'package:contexta/domain/repository/settings_repository.dart';
import 'package:contexta/domain/time/time_provider.dart';
import 'package:contexta/domain/usecase/startup_orchestration_usecase.dart';

/// 计划 B Task 5：启动编排（服务端同步模型）测试。
///
/// 语义（简报裁定）：
/// 1. 未 onboarding → StartupNeedsOnboarding（不触发同步）
/// 2. 已 onboarding + 无 token → StartupNeedsLogin（同步跳过，首页仍可用）
/// 3. 已 onboarding + 有 token → 同步执行 → 今天批次存在 → assignBatchForToday
///    （dailyCountSnapshot = settings.dailyArticleCount）→ Ready(syncedBatches: N)
/// 4. 已分配过（daily_learning 存在）→ 不重复分配
/// 5. 同步失败 → Ready(syncedBatches: 0)（不阻塞首页——历史文章可读）

/// ArticleRepository 桩：仅实现编排用到的 3 个方法，其余 noSuchMethod 兜底。
class FakeArticleRepository implements ArticleRepository {
  /// 今天已分配的批次（getAssignedBatchForDate 返回）。
  ArticleBatch? assignedToday;

  /// (难度, 今天) 批次（getBatchByDifficultyAndDate 返回）。
  ArticleBatch? todayBatch;

  final List<String> assignCalls = [];

  @override
  Future<ArticleBatch?> getAssignedBatchForDate(String readDate) async =>
      assignedToday;

  @override
  Future<ArticleBatch?> getBatchByDifficultyAndDate(
    String difficulty,
    String date,
  ) async => todayBatch;

  @override
  Future<bool> assignBatchForToday(
    int batchId,
    String refBatchDate,
    int dailyCount,
  ) async {
    assignCalls.add('$batchId:$refBatchDate:$dailyCount');
    return true;
  }

  @override
  dynamic noSuchMethod(Invocation invocation) => Future.value(null);
}

class FakeSettingsRepository implements SettingsRepository {
  UserSettings? settings = const UserSettings(
    isOnboarded: true,
    difficultyLevel: 'LOW',
    dailyArticleCount: 5,
  );

  @override
  Future<UserSettings?> getSettings() async => settings;

  @override
  Stream<UserSettings?> observeSettings() => const Stream.empty();

  @override
  dynamic noSuchMethod(Invocation invocation) => Future.value(null);
}

/// SyncArticlesUseCase 桩：记录调用次数，可配置结果 / 抛错。
class FakeSyncArticles extends SyncArticlesUseCase {
  FakeSyncArticles({
    this.result = const SyncResult(
      syncedBatches: 2,
      syncedArticles: 10,
      skippedAuth: false,
    ),
    this.throwError,
  }) : super(
         db: _db,
         batchDao: ArticleBatchDao(_db),
         articleDao: ArticleDao(_db),
         paragraphDao: ArticleParagraphDao(_db),
         fetchToday: () async => const [],
         timeProvider: const FakeTimeProvider(),
       );

  /// 只读的内存库（SyncArticlesUseCase 构造需要，本 fake 的 call 被覆写不走库）。
  static final _db = AppDatabase.forTesting(NativeDatabase.memory());

  final SyncResult result;
  Object? throwError;
  int calls = 0;

  @override
  Future<SyncResult> call() async {
    calls++;
    final e = throwError;
    if (e != null) throw e;
    return result;
  }
}

ArticleBatch _batch(
  int id,
  BatchStatus status, {
  String generatedOn = '2026-08-01',
}) => ArticleBatch(
  id: id,
  status: status,
  difficultyLevelSnapshot: 'LOW',
  generatedOn: generatedOn,
  lastUpdatedAt: '2026-08-01T21:25:42+08:00',
);

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  late FakeArticleRepository repo;
  late FakeSettingsRepository settings;
  late FakeSyncArticles sync;
  late StartupOrchestrationUseCase useCase;

  setUp(() {
    repo = FakeArticleRepository();
    settings = FakeSettingsRepository();
    sync = FakeSyncArticles();
    useCase = StartupOrchestrationUseCase(
      articleRepository: repo,
      settingsRepository: settings,
      timeProvider: const FakeTimeProvider(),
      syncArticles: sync,
    );
  });

  // ─── 1. 未 onboarding → NeedsOnboarding（不触发同步） ─────────────────

  test('未 onboarding 返回 NeedsOnboarding 且不触发同步', () async {
    settings.settings = const UserSettings(
      isOnboarded: false,
      difficultyLevel: 'LOW',
      dailyArticleCount: 5,
      serverToken: 'tok',
    );

    final result = await useCase();

    expect(result, isA<StartupNeedsOnboarding>());
    expect(sync.calls, 0);
    expect(repo.assignCalls, isEmpty);
  });

  // ─── 2. 已 onboarding + 无 token → NeedsLogin（同步跳过） ─────────────

  test('无 token 返回 NeedsLogin 且同步跳过', () async {
    settings.settings = const UserSettings(
      isOnboarded: true,
      difficultyLevel: 'LOW',
      dailyArticleCount: 5,
      serverToken: null,
    );

    final result = await useCase();

    expect(result, isA<StartupNeedsLogin>());
    expect(sync.calls, 0);
    expect(repo.assignCalls, isEmpty);
  });

  test('token 为空字符串同样视为未登录', () async {
    settings.settings = const UserSettings(
      isOnboarded: true,
      difficultyLevel: 'LOW',
      dailyArticleCount: 5,
      serverToken: '',
    );

    final result = await useCase();

    expect(result, isA<StartupNeedsLogin>());
    expect(sync.calls, 0);
  });

  // ─── 3. 有 token → 同步 → 分配今天批次 → Ready ──────────────────────

  test(
    '同步成功 + 今天批次存在 → 分配给今天（dailyCountSnapshot=设置值）→ Ready(syncedBatches: N)',
    () async {
      settings.settings = const UserSettings(
        isOnboarded: true,
        difficultyLevel: 'LOW',
        dailyArticleCount: 5,
        serverToken: 'tok',
      );
      repo.todayBatch = _batch(
        3,
        BatchStatus.current,
        generatedOn: '2026-08-01',
      );

      final result = await useCase();

      expect(result, isA<StartupReady>());
      expect((result as StartupReady).syncedBatches, 2);
      expect(sync.calls, 1);
      expect(repo.assignCalls, ['3:2026-08-01:5']);
    },
  );

  test('同步成功但今天无对应难度的批次 → 不分配，仍 Ready', () async {
    settings.settings = const UserSettings(
      isOnboarded: true,
      difficultyLevel: 'LOW',
      dailyArticleCount: 5,
      serverToken: 'tok',
    );

    final result = await useCase();

    expect(result, isA<StartupReady>());
    expect((result as StartupReady).syncedBatches, 2);
    expect(repo.assignCalls, isEmpty);
  });

  // ─── 4. 今天已分配 → 不重复分配 ─────────────────────────────────────

  test('今天已有 daily_learning → 不重复分配，仍 Ready', () async {
    settings.settings = const UserSettings(
      isOnboarded: true,
      difficultyLevel: 'LOW',
      dailyArticleCount: 5,
      serverToken: 'tok',
    );
    repo.assignedToday = _batch(
      1,
      BatchStatus.ready,
      generatedOn: '2026-07-31',
    );
    repo.todayBatch = _batch(3, BatchStatus.current, generatedOn: '2026-08-01');

    final result = await useCase();

    expect(result, isA<StartupReady>());
    expect((result as StartupReady).syncedBatches, 2);
    expect(sync.calls, 1);
    expect(repo.assignCalls, isEmpty);
  });

  // ─── 5. 同步失败 → Ready(0)（不阻塞首页） ───────────────────────────

  test('同步失败 → Ready(syncedBatches: 0)，不阻塞首页', () async {
    settings.settings = const UserSettings(
      isOnboarded: true,
      difficultyLevel: 'LOW',
      dailyArticleCount: 5,
      serverToken: 'tok',
    );
    sync.throwError = StateError('network down');
    repo.todayBatch = _batch(3, BatchStatus.current, generatedOn: '2026-08-01');

    final result = await useCase();

    expect(result, isA<StartupReady>());
    expect((result as StartupReady).syncedBatches, 0);
    // 同步失败直接降级返回，不做分配
    expect(repo.assignCalls, isEmpty);
  });
}

class FakeTimeProvider implements TimeProvider {
  const FakeTimeProvider();

  @override
  int nowMillis() => 1785636000000;

  @override
  String nowDateTimeString() => '2026-08-01T12:00:00+08:00';

  @override
  String todayDateString() => '2026-08-01';

  @override
  String nextDateString() => '2026-08-02';
}
